File: app/src/main/java/com/nexus/browser/ScraperEngine.kt
package com.nexus.browser

import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Locale
import java.util.regex.Pattern

/**
 * Public-resource scraper. It inspects the delivered HTML only; it does not bypass
 * authentication, CAPTCHA, DRM or access controls. Designed to be deterministic and bounded.
 */
data class ScrapeResult(
    val title: String,
    val links: List<String>,
    val media: List<String>,
    val htmlSize: Int
)

object ScraperEngine {
    private const val MAX_HTML_BYTES = 5 * 1024 * 1024
    private const val MAX_LINKS = 2500
    private const val MAX_MEDIA = 1000

    private val attrPattern = Pattern.compile(
        "(?:href|src|data-src|data-url|data-video|data-file|content)\\s*=\\s*['\\\"]([^'\\\"]+)['\\\"]",
        Pattern.CASE_INSENSITIVE
    )
    private val srcSetPattern = Pattern.compile("(?:srcset|data-srcset)\\s*=\\s*['\\\"]([^'\\\"]+)['\\\"]", Pattern.CASE_INSENSITIVE)
    private val titlePattern = Pattern.compile("<title[^>]*>(.*?)</title>", Pattern.CASE_INSENSITIVE or Pattern.DOTALL)
    private val ogPattern = Pattern.compile("<meta[^>]+(?:property|name)\\s*=\\s*['\\\"](?:og:video(?::secure_url)?|og:image|twitter:image|twitter:player:stream)['\\\"][^>]+content\\s*=\\s*['\\\"]([^'\\\"]+)['\\\"]", Pattern.CASE_INSENSITIVE)
    private val urlPattern = Pattern.compile("https?://[^\\s\\\"'<>\\\\]+", Pattern.CASE_INSENSITIVE)
    private val escapedUrlPattern = Pattern.compile("https?:\\\\?/\\\\?/[^\\s\\\"'<>]+", Pattern.CASE_INSENSITIVE)
    private val cssUrlPattern = Pattern.compile("url\\(\\s*['\\\"]?([^'\\\")]+)['\\\"]?\\s*\\)", Pattern.CASE_INSENSITIVE)

    fun scrape(pageUrl: String): ScrapeResult {
        require(pageUrl.startsWith("http://") || pageUrl.startsWith("https://"))
        val response = fetch(pageUrl)
        val html = response.body
        val base = URI(response.finalUrl)
        val links = LinkedHashSet<String>()
        val media = LinkedHashSet<String>()

        fun add(raw: String?) {
            if (raw.isNullOrBlank() || links.size >= MAX_LINKS) return
            val value = decodeHtml(raw.trim())
            if (value.startsWith("data:") || value.startsWith("javascript:") || value.startsWith("mailto:")) return
            val resolved = runCatching { base.resolve(value).toString() }.getOrNull() ?: return
            if (!resolved.startsWith("http://") && !resolved.startsWith("https://")) return
            links.add(resolved)
            if (isMedia(resolved) && media.size < MAX_MEDIA) media.add(resolved)
        }

        val attrs = attrPattern.matcher(html)
        while (attrs.find()) add(attrs.group(1))

        val srcsets = srcSetPattern.matcher(html)
        while (srcsets.find()) srcsets.group(1).split(',').forEach { add(it.trim().substringBefore(' ')) }

        val og = ogPattern.matcher(html)
        while (og.find()) add(og.group(1))

        val css = cssUrlPattern.matcher(html)
        while (css.find()) add(css.group(1))

        val rawUrls = urlPattern.matcher(html)
        while (rawUrls.find()) add(rawUrls.group())

        val escaped = escapedUrlPattern.matcher(html)
        while (escaped.find()) add(escaped.group().replace("\\/", "/"))

        // Media-looking JSON/config strings are common in JS players.
        val decoded = runCatching { URLDecoder.decode(html, StandardCharsets.UTF_8.name()) }.getOrDefault(html)
        val decodedUrls = urlPattern.matcher(decoded)
        while (decodedUrls.find()) add(decodedUrls.group())

        val finalLinks = links.take(MAX_LINKS)
        val finalMedia = media.filter { isMedia(it) }.distinct().take(MAX_MEDIA)
        val title = titlePattern.matcher(html).let {
            if (it.find()) decodeHtml(it.group(1)).replace(Regex("\\s+"), " ").trim().take(200)
            else base.host ?: "NEXUS"
        }
        return ScrapeResult(title, finalLinks, finalMedia, html.toByteArray(StandardCharsets.UTF_8).size)
    }

    private data class Response(val body: String, val finalUrl: String)

    private fun fetch(pageUrl: String): Response {
        val c = URL(pageUrl).openConnection() as HttpURLConnection
        c.connectTimeout = 12_000
        c.readTimeout = 20_000
        c.instanceFollowRedirects = true
        c.setRequestProperty("User-Agent", "Mozilla/5.0 (Android; NEXUS Browser/1.0)")
        c.setRequestProperty("Accept", "text/html,application/xhtml+xml,application/json,text/plain,*/*;q=0.8")
        c.setRequestProperty("Accept-Encoding", "gzip")
        return try {
            val bytes = c.inputStream.use { it.readNBytes(MAX_HTML_BYTES) }
            Response(bytes.toString(StandardCharsets.UTF_8), c.url.toString())
        } finally { c.disconnect() }
    }

    private fun isMedia(url: String): Boolean {
        val p = runCatching { URI(url).path.orEmpty().lowercase(Locale.US) }.getOrDefault("")
        return p.endsWith(".mp4") || p.endsWith(".m4v") || p.endsWith(".webm") || p.endsWith(".mov") ||
            p.endsWith(".m3u8") || p.endsWith(".mp3") || p.endsWith(".m4a") || p.endsWith(".aac") ||
            p.endsWith(".flac") || p.endsWith(".wav") || p.endsWith(".jpg") || p.endsWith(".jpeg") ||
            p.endsWith(".png") || p.endsWith(".webp") || p.endsWith(".gif") || p.endsWith(".avif")
    }

    private fun decodeHtml(value: String): String = value
        .replace("&amp;", "&", true)
        .replace("&quot;", "\"", true)
        .replace("&#39;", "'", true)
        .replace("&lt;", "<", true)
        .replace("&gt;", ">", true)
}

File: lib/engines/scraper_engine.dart
import 'dart:convert';
import 'dart:io';

import '../models/scrape_result.dart';

/// Public-resource scraper. Inspects only the delivered HTML; does not
/// bypass authentication, CAPTCHA, DRM or access controls. Deterministic
/// and bounded — 1:1-Verhalten zu Kotlins `ScraperEngine`.
///
/// Ein Unterschied zum Original ist erwähnenswert: `dart:io`s [HttpClient]
/// entpackt gzip-komprimierte Antworten standardmäßig automatisch
/// (`autoUncompress` ist per Default `true`). Der Bug, den wir im
/// Kotlin-Original erst nachträglich fixen mussten (gzip angefragt, aber
/// nie entpackt), kann hier grundsätzlich gar nicht erst auftreten.
class ScraperEngine {
  ScraperEngine._();

  static const int _maxHtmlBytes = 5 * 1024 * 1024;
  static const int _maxLinks = 2500;
  static const int _maxMedia = 1000;
  static const String _userAgent = 'Mozilla/5.0 (Linux; NEXUS Browser/1.0)';

  static final RegExp _attrPattern = RegExp(
    r'''(?:href|src|data-src|data-url|data-video|data-file|content)\s*=\s*['"]([^'"]+)['"]''',
    caseSensitive: false,
  );
  static final RegExp _srcSetPattern = RegExp(
    r'''(?:srcset|data-srcset)\s*=\s*['"]([^'"]+)['"]''',
    caseSensitive: false,
  );
  static final RegExp _titlePattern = RegExp(
    r'<title[^>]*>(.*?)</title>',
    caseSensitive: false,
    dotAll: true,
  );
  static final RegExp _ogPattern = RegExp(
    r'''<meta[^>]+(?:property|name)\s*=\s*['"](?:og:video(?::secure_url)?|og:image|twitter:image|twitter:player:stream)['"][^>]+content\s*=\s*['"]([^'"]+)['"]''',
    caseSensitive: false,
  );
  static final RegExp _urlPattern = RegExp(
    r'''https?://[^\s"'<>\\]+''',
    caseSensitive: false,
  );
  static final RegExp _escapedUrlPattern = RegExp(
    r'''https?:\\?/\\?/[^\s"'<>]+''',
    caseSensitive: false,
  );
  static final RegExp _cssUrlPattern = RegExp(
    r'''url\(\s*['"]?([^'")]+)['"]?\s*\)''',
    caseSensitive: false,
  );

  static Future<ScrapeResult> scrape(String pageUrl) async {
    if (!pageUrl.startsWith('http://') && !pageUrl.startsWith('https://')) {
      throw ArgumentError('pageUrl muss http(s) sein: $pageUrl');
    }
    final response = await _fetch(pageUrl);
    final html = response.body;
    final base = Uri.parse(response.finalUrl);
    final links = <String>{};
    final media = <String>{};

    void add(String? raw) {
      if (raw == null || raw.trim().isEmpty || links.length >= _maxLinks) {
        return;
      }
      final value = _decodeHtml(raw.trim());
      if (value.startsWith('data:') ||
          value.startsWith('javascript:') ||
          value.startsWith('mailto:')) {
        return;
      }
      Uri resolved;
      try {
        resolved = base.resolve(value);
      } catch (_) {
        return;
      }
      final resolvedStr = resolved.toString();
      if (!resolvedStr.startsWith('http://') &&
          !resolvedStr.startsWith('https://')) {
        return;
      }
      links.add(resolvedStr);
      if (_isMedia(resolvedStr) && media.length < _maxMedia) {
        media.add(resolvedStr);
      }
    }

    for (final m in _attrPattern.allMatches(html)) {
      add(m.group(1));
    }
    for (final m in _srcSetPattern.allMatches(html)) {
      for (final part in (m.group(1) ?? '').split(',')) {
        add(part.trim().split(' ').first);
      }
    }
    for (final m in _ogPattern.allMatches(html)) {
      add(m.group(1));
    }
    for (final m in _cssUrlPattern.allMatches(html)) {
      add(m.group(1));
    }
    for (final m in _urlPattern.allMatches(html)) {
      add(m.group(0));
    }
    for (final m in _escapedUrlPattern.allMatches(html)) {
      add(m.group(0)?.replaceAll(r'\/', '/'));
    }

    // Media-looking JSON/config strings are common in JS players.
    String decoded;
    try {
      decoded = Uri.decodeFull(html);
    } catch (_) {
      decoded = html;
    }
    for (final m in _urlPattern.allMatches(decoded)) {
      add(m.group(0));
    }

    final finalLinks = links.take(_maxLinks).toList();
    final finalMedia =
        media.where(_isMedia).toSet().take(_maxMedia).toList();

    final titleMatch = _titlePattern.firstMatch(html);
    final title = titleMatch != null
        ? _decodeHtml(titleMatch.group(1) ?? '')
            .replaceAll(RegExp(r'\s+'), ' ')
            .trim()
            .let((s) => s.length > 200 ? s.substring(0, 200) : s)
        : (base.host.isNotEmpty ? base.host : 'NEXUS');

    return ScrapeResult(
      title: title,
      links: finalLinks,
      media: finalMedia,
      htmlSize: utf8.encode(html).length,
    );
  }

  static Future<_Response> _fetch(String pageUrl) async {
    final client = HttpClient()
      ..connectionTimeout = const Duration(seconds: 12)
      ..autoUncompress = true;
    try {
      final request = await client
          .getUrl(Uri.parse(pageUrl))
          .timeout(const Duration(seconds: 12));
      request.headers.set(HttpHeaders.userAgentHeader, _userAgent);
      request.headers.set(
        HttpHeaders.acceptHeader,
        'text/html,application/xhtml+xml,application/json,text/plain,*/*;q=0.8',
      );
      final response =
          await request.close().timeout(const Duration(seconds: 20));
      final bytes = <int>[];
      await for (final chunk in response) {
        bytes.addAll(chunk);
        if (bytes.length >= _maxHtmlBytes) break;
      }
      final capped = bytes.length > _maxHtmlBytes
          ? bytes.sublist(0, _maxHtmlBytes)
          : bytes;
      return _Response(
        body: utf8.decode(capped, allowMalformed: true),
        finalUrl: response.redirects.isNotEmpty
            ? response.redirects.last.location.toString()
            : pageUrl,
      );
    } finally {
      client.close(force: true);
    }
  }

  static bool _isMedia(String url) {
    String path;
    try {
      path = Uri.parse(url).path.toLowerCase();
    } catch (_) {
      path = '';
    }
    const extensions = [
      '.mp4', '.m4v', '.webm', '.mov', '.m3u8', '.mp3', '.m4a', '.aac',
      '.flac', '.wav', '.jpg', '.jpeg', '.png', '.webp', '.gif', '.avif', //
    ];
    return extensions.any(path.endsWith);
  }

  static String _decodeHtml(String value) => value
      .replaceAll('&amp;', '&')
      .replaceAll('&quot;', '"')
      .replaceAll('&#39;', "'")
      .replaceAll('&lt;', '<')
      .replaceAll('&gt;', '>');
}

class _Response {
  final String body;
  final String finalUrl;
  const _Response({required this.body, required this.finalUrl});
}

extension _Let<T> on T {
  R let<R>(R Function(T) block) => block(this);
}
